### ISSUE TITLE
integration flake: integration.TestPersistentVolumeRecycler test flaky (timed out after 5 minutes)

### ISSUE THREAD

**Issue by yujuhong** [kubernetes/kubernetes#18050]
Relevant log snippet:

Failure example: https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/de79267dc310878fe4d95e70f9878a3ebdab861a/kubernetes-pull-test-unit-integration/3977/

```
17:21:00 E1202 01:21:00.110624   22517 persistentvolume_claim_binder_controller.go:146] PVClaimBinder could not update claim fake-pvc: Unexpected error saving claim status: persistentvolumeclaims "fake-pvc" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
17:21:02 W1202 01:21:02.136975   22517 controller.go:260] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[]} [{[{192.168.10.4 <nil>}] [] [{https 6443 TCP}]}]}
17:21:02 E1202 01:21:02.206620   22517 etcd_watcher.go:356] failure to decode api object: '' from &etcd.Response{Action:"delete", Node:(*etcd.Node)(0xc208ed63c0), PrevNode:(*etcd.Node)(0xc208ed6420), EtcdIndex:0xb5, RaftIndex:0xdc, RaftTerm:0x2} &etcd.Node{Key:"/registry", Value:"", Dir:true, Expiration:(*time.Time)(nil), TTL:0, Nodes:etcd.Nodes(nil), ModifiedIndex:0xad, CreatedIndex:0xad}
17:21:02 E1202 01:21:02.206913   22517 etcd_watcher.go:356] failure to decode api object: '' from &etcd.Response{Action:"delete", Node:(*etcd.Node)(0xc208ed6660), PrevNode:(*etcd.Node)(0xc208ed66c0), EtcdIndex:0xbb, RaftIndex:0xe4, RaftTerm:0x2} &etcd.Node{Key:"/registry", Value:"", Dir:true, Expiration:(*time.Time)(nil), TTL:0, Nodes:etcd.Nodes(nil), ModifiedIndex:0xad, CreatedIndex:0xad}
17:21:02 E1202 01:21:02.207374   22517 etcd_watcher.go:356] failure to decode api object: '' from &etcd.Response{Action:"delete", Node:(*etcd.Node)(0xc208ed69c0), PrevNode:(*etcd.Node)(0xc208ed6a20), EtcdIndex:0xbb, RaftIndex:0xe4, RaftTerm:0x2} &etcd.Node{Key:"/registry", Value:"", Dir:true, Expiration:(*time.Time)(nil), TTL:0, Nodes:etcd.Nodes(nil), ModifiedIndex:0xad, CreatedIndex:0xad}
17:21:02 E1202 01:21:02.207571   22517 etcd_watcher.go:356] failure to decode api object: '' from &etcd.Response{Action:"delete", Node:(*etcd.Node)(0xc208ed6c60), PrevNode:(*etcd.Node)(0xc208ed6cc0), EtcdIndex:0xbb, RaftIndex:0xe4, RaftTerm:0x2} &etcd.Node{Key:"/registry", Value:"", Dir:true, Expiration:(*time.Time)(nil), TTL:0, Nodes:etcd.Nodes(nil), ModifiedIndex:0xad, CreatedIndex:0xad}
17:21:03 E1202 01:21:03.453637   22517 persistentvolume_claim_binder_controller.go:146] PVClaimBinder could not update claim fake-pvc: Unexpected error saving claim status: persistentvolumeclaims "fake-pvc" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
17:21:06 W1202 01:21:06.935467   22517 controller.go:260] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[]} [{[{192.168.10.4 <nil>}] [] [{https 6443 TCP}]}]}
17:21:06 W1202 01:21:06.935815   22517 controller.go:260] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[]} [{[{192.168.10.4 <nil>}] [] [{https 6443 TCP}]}]}
17:21:06 W1202 01:21:06.940809   22517 controller.go:260] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[]} [{[{192.168.10.4 <nil>}] [] [{https 6443 TCP}]}]}
17:21:06 E1202 01:21:06.979803   22517 controller.go:106] unable to sync kubernetes service: Endpoints "kubernetes" is invalid: [uid: invalid value 'f64acd48-9892-11e5-9060-0242ac110002', Details: field is immutable, metadata.uid: invalid value 'f64acd48-9892-11e5-9060-0242ac110002', Details: field is immutable]
17:21:06 E1202 01:21:06.980073   22517 controller.go:106] unable to sync kubernetes service: Endpoints "kubernetes" is invalid: [uid: invalid value 'f64a19c0-9892-11e5-9060-0242ac110002', Details: field is immutable, metadata.uid: invalid value 'f64a19c0-9892-11e5-9060-0242ac110002', Details: field is immutable]
17:21:07 W1202 01:21:07.561835   22517 controller.go:295] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default  f64a0364-9892-11e5-9060-0242ac110002 201 0 2015-12-02 01:21:06 +0000 UTC <nil> <nil> map[] map[]} [{[{<nil> <nil>}] [] [{https 6443 TCP}]}]}
17:21:07 E1202 01:21:07.562522   22517 controller.go:106] unable to sync kubernetes service: Endpoints "kubernetes" is invalid: [subsets[0].addresses[0].ip: invalid value '<nil>', Details: invalid IPv4 address, subsets[0].addresses[0].ip: invalid value '<nil>', Details: invalid IPv4 address]
17:21:17 W1202 01:21:17.564571   22517 controller.go:295] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default  f64a0364-9892-11e5-9060-0242ac110002 201 0 2015-12-02 01:21:06 +0000 UTC <nil> <nil> map[] map[]} [{[{<nil> <nil>}] [] [{https 6443 TCP}]}]}
...
<omitted similar messages>
...
panic: test timed out after 5m0s
```

/cc @markturansky, who added the test.

**Comment by markturansky** [comment]
@yujuhong how frequently do you see this timeout?  What version of Kube and Go?

I have not been able to reproduce the timeout.

**Comment by yujuhong** [comment]
I saw the timeout on two different PR builds yesterday. Just looked the recent build failures this morning, and there are two more:
http://kubekins.dls.corp.google.com:8081/job/kubernetes-pull-test-unit-integration/4059/
http://kubekins.dls.corp.google.com:8081/job/kubernetes-pull-test-unit-integration/4047/

Maybe @kubernetes/goog-testing can shed some light on the testing environment of the per-PR unit tests.

**Comment by markturansky** [comment]
I don't have access those those internal sites.

I will test further.

**Comment by ixdy** [comment]
On the PR builds we're running with Ginkgo in parallel mode, so it's possible that this test is interacting poorly with other tests running simultaneously?

**Comment by markturansky** [comment]
Ah, that might be.

I don't think integration tests are inherently parallelizable.  Many of the integration tests call `framework.DeleteAllEtcdKeys()`, which means that one test will clobber another's data.  If a test is waiting on some condition in that data (and the PV integration test does), it will wait until it times out.  The data will never be in that state.  It's gone.

https://github.com/kubernetes/kubernetes/blob/master/test/integration/client_test.go#L47

**Comment by gmarek** [comment]
I just saw a similar failure while running from integration tests locally:
```
E1203 09:59:25.691273    5321 persistentvolume_claim_binder_controller.go:146] PVClaimBinder could not update claim fake-pvc: Unexpected error saving claim status: persistentvolumeclaims "fake-pvc" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
W1203 09:59:26.806404    5321 controller.go:260] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[]} [{[{192.168.10.4 <nil>}] [] [{https 6443 TCP}]}]}
E1203 09:59:27.264903    5321 etcd_watcher.go:356] failure to decode api object: '' from &etcd.Response{Action:"delete", Node:(*etcd.Node)(0xc8203b9200), PrevNode:(*etcd.Node)(0xc8203b9260), EtcdIndex:0x265, RaftIndex:0x3a1, RaftTerm:0x2} &etcd.Node{Key:"/kubernetes.io", Value:"", Dir:true, Expiration:(*time.Time)(nil), TTL:0, Nodes:etcd.Nodes(nil), ModifiedIndex:0x256, CreatedIndex:0x256}
E1203 09:59:27.265043    5321 etcd_watcher.go:356] failure to decode api object: '' from &etcd.Response{Action:"delete", Node:(*etcd.Node)(0xc820ebbe60), PrevNode:(*etcd.Node)(0xc820ebbec0), EtcdIndex:0x265, RaftIndex:0x3a1, RaftTerm:0x2} &etcd.Node{Key:"/kubernetes.io", Value:"", Dir:true, Expiration:(*time.Time)(nil), TTL:0, Nodes:etcd.Nodes(nil), ModifiedIndex:0x256, CreatedIndex:0x256}
E1203 09:59:27.265118    5321 etcd_watcher.go:356] failure to decode api object: '' from &etcd.Response{Action:"delete", Node:(*etcd.Node)(0xc8207d1140), PrevNode:(*etcd.Node)(0xc8207d11a0), EtcdIndex:0x25e, RaftIndex:0x398, RaftTerm:0x2} &etcd.Node{Key:"/kubernetes.io", Value:"", Dir:true, Expiration:(*time.Time)(nil), TTL:0, Nodes:etcd.Nodes(nil), ModifiedIndex:0x256, CreatedIndex:0x256}
E1203 09:59:27.265123    5321 etcd_watcher.go:356] failure to decode api object: '' from &etcd.Response{Action:"delete", Node:(*etcd.Node)(0xc82053df80), PrevNode:(*etcd.Node)(0xc820adc000), EtcdIndex:0x265, RaftIndex:0x3a1, RaftTerm:0x2} &etcd.Node{Key:"/kubernetes.io", Value:"", Dir:true, Expiration:(*time.Time)(nil), TTL:0, Nodes:etcd.Nodes(nil), ModifiedIndex:0x256, CreatedIndex:0x256}
W1203 09:59:28.090660    5321 controller.go:260] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[]} [{[{192.168.10.4 <nil>}] [] [{https 6443 TCP}]}]}
E1203 09:59:28.600235    5321 persistentvolume_claim_binder_controller.go:146] PVClaimBinder could not update claim fake-pvc: Unexpected error saving claim status: persistentvolumeclaims "fake-pvc" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
panic: test timed out after 4m0s
```

**Comment by wojtek-t** [comment]
@gmarek asked my about errors from "etcd_watcher.go" - those aren't dangerous. Those are the result of framework.DeleteAllEtcdKeys() - once you remove the whole directory from etcd, etcd doesn't send individual deleted objects to the watcher, but just the whole directory (without its elements). This is causing such errors to be logged. This isn't the problem here.

**Comment by markturansky** [comment]
@wojtek-t 

If Test A sets up Pod-foo and waits for some condition to be fulfilled, that condition will never happen if Test B calls DeleteAllEtcdKeys.

That, I believe, is the fundamental issue with integration tests running in parallel.

I would love to have someone tell me I'm wrong about that or that I'm doing something incorrect in my integration test.

**Comment by wojtek-t** [comment]
Sorry - I missed your previous post.

Yes - I agree that if there are multiple integrations tests running in parallel and they are writing to the same etcd instance, this will simply not work.

However - I'm not sure if this is the case for us - is it?

**Comment by markturansky** [comment]
I believe it is.  The PV integration test runs the various PV controllers.  PVs and PVCs are posted and then the test waits for the controllers to do their work.

If the data is gone, the test will time out.

**Comment by markturansky** [comment]
@wojtek-t 

> On the PR builds we're running with Ginkgo in parallel mode

So, yes, I believe DeleteAllEtcdKeys is the culprit.

It would be easy to test by disabling parallel mode in the builds.

**Comment by wojtek-t** [comment]
OK - then it definitely can be an issue (to be honest I'm surprised, that other tests work correctly...)

**Comment by markturansky** [comment]
Am I relying too much on waits for conditions to be fulfilled?

**Comment by wojtek-t** [comment]
I think the problem here is the common underlying etcd and the fact that every test is removing all contents of etcd at its beginning.

We should have a similar mechanism to what we have in e2e tests:
- each integration test is running in a single namespace
- each test is deleting that namespace on its end

@ixdy @quinton-hoole @kubernetes/goog-testing - thoughts?

**Comment by markturansky** [comment]
> each test is deleting that namespace on its end

Except for cluster objects, like nodes and PVs.

I don't mind purposefully cleaning up etcd after a test (that is, I delete all PVs I made).

**Comment by ixdy** [comment]
I miswrote earlier - I thought we were referring to e2e tests, not integration tests.

We may end up running several instances of the integration test simultaneously, but each instance is running in its own Docker container, so that shouldn't cause any problems.

**Comment by wojtek-t** [comment]
Yeah - that was what I actually thought. I was surprised that we are running integration tests in parallel.

**Comment by markturansky** [comment]
@ixdy @wojtek-t Where does that leave this P0 issue?  are you going to change the build to !parallel mode? 

Changing how the tests run (to be in separate container, separate etcd, ...) is a bit out of scope for me and the work I have to do.  Meanwhile, it's my test that is flaky.

**Comment by wojtek-t** [comment]
@markturansky - integration tests are NOT running in parallel. So this is not the reason of those failures (note that almost all tests would be failing in such case).

**Comment by markturansky** [comment]
I'm not familiar with Shippable's config file, but this looks like concurrent integration tests:  https://github.com/kubernetes/kubernetes/blob/master/shippable.yml#L65

**Comment by gmarek** [comment]
We're not running them on shippable, or at least we don't care about those runs. Both unit and integration tests are run on Jenkins.

**Comment by ixdy** [comment]
On Jenkins we run `./hack/jenkins/gotest-dockerized.sh`.

**Comment by gmarek** [comment]
@markturansky - any progress on this?

**Comment by markturansky** [comment]
@gmarek how frequently is this still occurring?  the sync period for the pv controllers in the tests was dramatically shortened recently.

**Comment by gmarek** [comment]
I don't know, I'm not monitoring failures very closely - I believe I saw this on one of per-push tests.

**Comment by thockin** [comment]
We're making a surge to get test flakes taken care of - they are KILLING merge throughput.  Mark, do you think the logs are sufficient to catch this if it flakes again?  If not, what can we add to the logs to make forward progress?

**Comment by markturansky** [comment]
I think the issue was 10m sync on the controllers.    The correct state is eventually reached with enough sync periods, but the default 10m timeout is longer than the integration test timeout.

The test controllers have their sync periods down to 10s or so.  Is this particular timeout/flake still occurring with the shorter sync period?

**Comment by quinton-hoole** [comment]
@markturansky In which PR was the sync period dropped from 10 minutes to 10 seconds?  We can check whether there's any correlation between that merging and a drop in test failures.

**Comment by markturansky** [comment]
@quinton-hoole 

https://github.com/kubernetes/kubernetes/pull/14537

https://github.com/kubernetes/kubernetes/commit/4fc1bf1f2314b213af2f540ebe86d169ccfadd9c

**Comment by quinton-hoole** [comment]
@markturansky that PR merged on Oct 12, 2015.  This issue was reported on Dec 1, 2015.  Am I missing something?

**Comment by quinton-hoole** [comment]
@markturansky Sorry, I looked at the commit, not the PR.  The PR merged on Dec 11.  Let me check.

**Comment by markturansky** [comment]
@quinton-hoole this particular flake hit a few of my own PRs, but i haven't seen it since 14537 was merged.  I'm hoping that the sync period was culprit.   If not, please let me know what I can do to help.

**Comment by quinton-hoole** [comment]
@markturansky Our test history only goes back as far as Jan 17, but we haven't seen any failures since then (198 runs in total).  I think it's safe to close it.  We can re-open if we see it again.
